Conversion of LaTeX spec to asciidoc. (#53)
riscv/riscv-aia: src/AdvPLIC.adoc

numbers that are not _actual_ hart index numbers in the domain. For
example, the first IDC structure is always for hart index 0, but 0 is
not necessarily a valid index number for any hart in the domain. For
each IDC structure in the array that does not correspond to a valid hart
index number in the domain, the IDC structure's registers may (or may
not) be all read-only zeros.

Aside from the registers in
<<TableAdvPLIC-domainControlRegion>>
and those listed above for IDC structures, all other bytes in an
interrupt domain's control region are reserved and are implemented as
read-only zeros.

Only naturally aligned 32-bit simple reads and writes are supported
within an interrupt domain's control region. Writes to read-only bytes
are ignored. For other forms of accesses (other sizes, misaligned
accesses, or AMOs), implementations should preferably report an access
fault or bus error but must otherwise ignore the access.

The registers of the first 16 KiB of an interrupt domain's control
region (all but the IDC structures) are documented individually below.
IDC structures are documented later, in
<<AdvPLIC-directMode>>, "Interrupt delivery directly by

> the APLIC.""

Reviewer: ```suggestion
the APLIC."
```

choose, independently for each interrupt source, what other values are
supported for SM.

[[TableAdvPLIC-sourcecfg-SM]]
.Encoding of the SM (Source Mode) field of a sourcecfg register when bit D = 0
[%autowidth,float="center",align="center",cols="^,^,<",options="header"]
|===
|Value |Name |Description
|0 |Inactive |Inactive in this domain (and not delegated)
|1 |Detached |Active, detached from the source wire
|2–3 |— |_Reserved_
|4 |Edge1 |Active, edge-sensitive; interrupt asserted on rising edge
|5 |Edge0 |Active, edge-sensitive; interrupt asserted on falling edge
|6 |Level1 |Active, level-sensitive; interrupt asserted when high
|7 |Level0 |Active, level-sensitive; interrupt asserted when low
|===

An interrupt source is inactive in the interrupt domain if either the
source is delegated to a child domain (D = 1) or it is not delegated
(D = 0) and SM is Inactive. Whenever interrupt source latexmath:[$i$] is
inactive in an interrupt domain, the corresponding interrupt-pending and
interrupt-enable bits within the domain are read-only zeros, and
register `target[latexmath:[$i$]]` is also read-only zero. If source latexmath:[$i$] is changed

> from inactive to an active mode, the interrupt source 's pending and

Reviewer: ```suggestion
from inactive to an active mode, the interrupt source's pending and
```

enable bits remain zeros, unless set automatically for a reason
specified later in this section or in
<<AdvPLIC-pendingBits>>, and the defined subfields of `target[latexmath:[$i$]]` obtain UNSPECIFIED values.

When a source is configured as Detached, its wire input is ignored;
however, the interrupt-pending bit may still be set by a write to a `setip` or `setipnum` register. (This mode can be useful for receiving MSIs, for example.)

An edge-sensitive source can be configured to recognize an incoming
interrupt on either a rising edge (low-to-high transition) or a falling
edge (high-to-low transition). When configured for a falling edge (mode
Edge0), the source is said to be _inverted_.

A level-sensitive source can be configured to interpret either a high
level (1) or a low level (0) on the wire as the assertion of an
interrupt. When configured for a low level (mode Level0), the source is
said to be _inverted_.

For an interrupt source that is configured as edge-sensitive or
level-sensitive, define

> _rectified input value_ = (incoming wire value) XOR (source is

Reviewer: ```suggestion
----
_rectified input value_ = (incoming wire value) XOR (source is
```

> inverted).

Reviewer: ```suggestion
inverted).
----
```

to zero (direct delivery mode) will have the same effect as setting IE
to zero. See <<IMSIC-reg-eidelivery>>
and <<AdvPLIC-directMode-intrDelivery>>.

BE (Big-Endian) is a *WARL* field that determines the byte order for most
registers in the interrupt domain's memory-mapped control region. If
BE = 0, byte order is little-endian, and if BE = 1, it is big-endian.
For RISC-V systems that support only little-endian, BE may be read-only zero,
and for those that support only big-endian, BE may be read-only one. For
bi-endian systems, BE is writable.

Field BE affects the byte order of accesses to the `domaincfg` register itself, just
as for other registers in the interrupt domain’s control region. To deal
with this fact, the read-only value in `domaincfg’s` most-significant byte, bits
31:24, serves two purposes. First, for any read of `domaincfg`, the register's correct byte order is easily determined from the four-byte value
obtained: When interpreted in the correct byte order, bit 31 is one, and
in the wrong order, bit 31 is zero. Second, if the value of BE is
uncertain (prior to software initializing the interrupt domain,

> presumably), an 8-bit value latexmath:[$x$] can be safely written to `domaincfg` by writing (latexmath:[$x$]  24) latexmath:[$x$], where <<24 represents

Author: Fixed.

enable bits remain zeros, unless set automatically for a reason
specified later in this section or in
<<AdvPLIC-pendingBits>>, and the defined subfields of `target[latexmath:[$i$]]` obtain UNSPECIFIED values.

When a source is configured as Detached, its wire input is ignored;
however, the interrupt-pending bit may still be set by a write to a `setip` or `setipnum` register. (This mode can be useful for receiving MSIs, for example.)

An edge-sensitive source can be configured to recognize an incoming
interrupt on either a rising edge (low-to-high transition) or a falling
edge (high-to-low transition). When configured for a falling edge (mode
Edge0), the source is said to be _inverted_.

A level-sensitive source can be configured to interpret either a high
level (1) or a low level (0) on the wire as the assertion of an
interrupt. When configured for a low level (mode Level0), the source is
said to be _inverted_.

For an interrupt source that is configured as edge-sensitive or
level-sensitive, define

----
_rectified input value_ = (incoming wire value) XOR (source is
inverted).
----

For a source that is inactive or Detached, the _rectified input value_
is zero.

> Any write to a `sourcecfg` register might (or might not) cause the corresponding interrupt-pending bit to be set to one if the rectified input value is high (= 1) under the new source mode. A write to a `sourcescfg` register will not by itself cause a pending bit to be cleared except when the source is made inactive. (But see <<AdvPLIC-pendingBits>>.)

Author: Fixed

For systems built mainly for big-endian byte order, `setipnum_be` may be useful as a write port for MSIs from some devices.

[[AdvPLIC-reg-genmsi]]
==== Generate MSI (`genmsi`)

When the interrupt domain is configured in MSI delivery mode (`domaincfg`.DM = 1), register `genmsi` can be used to cause an _extempore_ MSI to be sent from the
APLIC to a hart. The main purpose for this function is to assist in
establishing a temporary known ordering between a hart's writes to the
APLIC's registers and the transmission of MSIs from the APLIC to the
hart, as explained later in <<AdvPLIC-MSISync>>.

[NOTE]
====
For other purposes, sending an MSI to a hart is usually better done by
writing directly to the hart's IMSIC, rather than employing an APLIC as
an intermediary. Use of the `genmsi` register should be minimized to avoid it
becoming a bottleneck.
====

Register `genmsi` has this format:
[%autowidth,float="center",align="center",cols="<,<",grid=none,frame=none]
|===
|bits 31:18 |Hart Index (*WLRL*)
|bits 12 |Busy (read-only)
|bits 10:0 |EIID (*WARL*)
|===

All other register bits are reserved and read as zeros.

> The Busy bit is ordinarily zero (false), but a write to genmsi` causes Busy to become one (true), indicating an extempore MSI is pending. The Hart

Author: Fixed

Index is the number of the target hart's guest interrupt file to which
MSIs will be sent. When Guest Index is zero, MSIs from a
supervisor-level domain are forwarded to the target hart at supervisor
level. For a machine-level domain, Guest Index is read-only zero, and
MSIs are forwarded to a target hart always at machine level.

Together, fields Hart Index and Guest Index of register `target[latexmath:[$i$]]` determine the
address for MSIs forwarded for interrupt source latexmath:[$i$]. The
remaining field EIID (External Interrupt Identity) specifies the data
value for those MSIs, eventually becoming the minor identity for an
external interrupt at the target hart.

If the interrupt domain's harts have IMSIC interrupt files that
implement latexmath:[$N$] distinct interrupt identities
(<<IMSIC-intrFilesAndIdents>>),
then EIID is a latexmath:[$k$]-bit unsigned integer field, where

> latexmath:[$\lceil\log_{2}N\rceil \leq k \leq \mbox{11}$]. EIID is thus

Author: Fixed

high. The pending bit cannot be set by a write to a `setip` or `setipnum` register.
* The pending bit is cleared whenever the rectified input value is low.
The pending bit is not cleared by a claim of the interrupt at the APLIC,
nor can it be cleared by a write to an `in_clrip` register or to `clripnum`.

If the source mode is Level1 or Level0 and the interrupt domain is
configured in MSI delivery mode (`domaincfg`.DM = 1):

* The pending bit is set to one by a low-to-high transition in the
rectified input value. The pending bit may also be set by a relevant
write to a `setip` or `setipnum` register when the rectified input value is high, but not when the rectified input value is low.
* The pending bit is cleared whenever the rectified input value is low,
when the interrupt is forwarded by MSI, or by a relevant write to an `in_clrip` register or to `clripnum`.

[NOTE]
====
When an interrupt domain is in direct delivery mode, the pending bit for
a level-sensitive source is always just a copy of the rectified input
value. Even in MSI delivery mode, the pending bit for a level-sensitive
source is never set (= 1) when the rectified input value is low.
====

> In addition to the rules above, a write to a register can cause the

Author: Fixed

==== Addresses and data for outgoing MSIs

To forward interrupts by MSIs, an APLIC must know the MSI target address
for each hart. For any given system, these addresses are fixed and
should be hardwired into the APLIC if possible. However, some APLIC
implementations may require that software supply the MSI target
addresses. In that case, the root domain's registers `mmsiaddrcfg`, `mmsiaddrcfgh`, `smsiaddrcfg`, and `smsiaddrcfgh` <<AdvPLIC-reg-mmsiaddrcfg>>
and <<AdvPLIC-reg-smsiaddrcfg>> may be used to configure the
MSI addresses for all interrupt domains. Alternatively MSI addresses may
be configured by some custom means outside this standard. If MSI target
addresses must be configured by software, this should be done only from
a suitably privileged execution mode, typically just once, early after
system reset.

For a machine-level interrupt domain, if MSI target addresses are
determined by `mmsiaddrcfg` and `mmsiaddrcfgh`, then the address for an outgoing MSI for interrupt
source latexmath:[$i$] is computed from those registers and from the
Hart Index field of register as follows:

[%autowidth,float="center",align="center",cols="<",grid=none,frame=none]
|===

> |latexmath:[$g = (Hart Index>>LHXW){\text{&}}({2}^{HHXW}-{1})$]

Author: Replaced latexmath with asciidoc.

> |latexmath:[$h = {Hart Index \text{&} } ({2}^{LHXW}-1)$]

Author: Replaced latexmath with asciidoc.

> |MSI address = (Base PPN \| (g<<(HHXS+12)) \| (h<<LHXS)<<12

Author: This particular formula doesn't have the Guest Index text as per the LaTeX version.
